File: irrgarten/src/irrgarten/Player.java
package irrgarten;

import java.util.ArrayList;

/**
 * Clase Player. Esta clase contiene la implementación de la variable player y sus 
 * funciones básicas para la utilización de la clase
 * 
 * @author Miguel Ángel Luque Gómez
 * correo: e.mangelluqg@go.ugr.es
 * 
 * @author Sergio Calvo González
 * correo: e.sergiocg10@go.ugr.es
 */ 
public class Player extends LabyrinthCharacter{
    private static final int INVALID_POS=-1;//Variable que indica la posición inicial del monstruo
    private static int MAX_WEAPONS = 2;     //Número de armas máximas que puede llevar el jugador
    private static int MAX_SHIELDS = 3;     //Número de escudos máximos que puede llevar el jugador
    private static int INITIAL_HEALTH = 3;  //Cantidad de vida con la que empieza el jugador
    private static int HITS2LOSE = 3;       //Cantidad de golpes recibidos para perder
    
    private char number;                    //Número de jugador
    private int consecutiveHits = 0;        //Golpes consecutivos recibidos por el jugador
    
    private ArrayList<Weapon> weapons= new ArrayList<>();   //Array donde almacenaremos las armas del jugador
    private ArrayList<Shield> shields= new ArrayList<>();   //Array donde almacenaremos los escudos del jugador
    
    private WeaponCardDeck weaponCardDeck;
    private ShieldCardDeck shieldCardDeck;
    
    /**
     * Constructor por parámetro de la clase Player.
     * 
     * @param nmb Número del jugador
     * @param howsmart Cómo de inteligente es
     * @param howstrong Cómo de fuerte es    
     */
    Player(char nmb, float howsmart, float howstrong){
        super("Player#"+nmb, howsmart, howstrong, (float)INITIAL_HEALTH);
        number = nmb;
        weaponCardDeck = new WeaponCardDeck();
        shieldCardDeck = new ShieldCardDeck();
    }
    
    //P5
    Player(Player other){
        super("Player#"+other.number, other.getIntelligence(), other.getStrength(), other.getHealth());
        this.setPos(other.getRow(), other.getCol());
        
        number = other.number;
        consecutiveHits = other.consecutiveHits;
        //Añadimos las armas
        for(int i = 0; i < other.weapons.size(); i++){
            this.weapons.add(other.weapons.get(i));
        }
        
        //Añadimos los escudos
        for(int i = 0; i < other.shields.size(); i++){
            this.shields.add(other.shields.get(i));
        }
    }
    
    /**
     * Funcion resurrect. Resucita al jugador, perdiendo las armas y escudos 
     * que poseía y recuperando la salud máxima, además de resetear el número de
     * golpes recibidos.
     */
    public void resurrect(){
        weapons.removeAll(weapons);
        shields.removeAll(shields);
        this.setHealth((float)INITIAL_HEALTH); 
        resetHits();       
    }
    
    /**
     * Funcion getNumber. Devuelve el número del jugador
     * 
     * @return Número del jugador 
     */
    public char getNumber(){
        return number;
    }
      
    /**
     * Mueve al personaje en una dirección, si esa dirección no es posible se mueve hacia una válida
     * @param direction Direccion preferente a moverse
     * @param validMoves Vector con direcciones posibles para moverse
     * @return direccion a la que se mueve finalmente
     */
    public Directions move(Directions direction, ArrayList<Directions> validMoves )
    {
        Directions move;
        int size=validMoves.size();
        boolean contained=false;
        for(int i=0;i<size && !contained;i++){
            if(validMoves.get(i)==direction){
                contained=true;
            }
        }
        if(size>0 && !contained)
        {
            Directions firstElement=validMoves.get(0);
            move=firstElement;
        }else
            move=direction;
        return move;
    }
    
    @Override
    /**
     * Funcion attack. Devuelve el valor de ataque del jugador, calculandolo
     * como la suma del ataque de sus armas y la suma de su fuerza
     * 
     * @return float con el valor de ataque
     */
    public float attack(){
        return this.getStrength() + sumWeapons();
    }
    
    @Override
    /**
     * Funcion defend. Devuelve si el jugador se defiende del ataque
     * @param receivedAttack Cantidad de daño del atacante
     * @return True si se defiende, false si no
     */
    public boolean defend(float receivedAttack){
        return manageHit(receivedAttack);
    }
    
    /**
     * Le da al jugador armas escudos y vida aleatoris como recompensa
     */
    public void receiveReward()
    {
        int wReward=Dice.weaponsReward();
        System.out.print("\t\tJugador "+this.number+" recibe "+wReward+" armas\n");
        int sReward=Dice.shieldsReward();
        System.out.print("\t\tJugador "+this.number+" recibe "+sReward+" escudos\n");

        for(int i=0;i<wReward;i++)
        {
            //Weapon wnew=this.newWeapon();
            this.receiveWeapon( this.weaponCardDeck.nextCard());
        }
        for(int i=0;i<sReward;i++)
        {
            //Shield snew=this.newShield();
            this.receiveShield( this.shieldCardDeck.nextCard());
        }
        int extraHealth=Dice.healthReward();
        this.setHealth(this.getHealth()+extraHealth);
        System.out.print("\t\tJugador "+this.number+" recibe "+this.getHealth()+" corazones\n");

    }
    
    @Override
    /**
     * Funcion toString. Devuelve la información del jugador, su vida, inteligencia fuerza
     * y la infomación de sus armas y escudos.
     * @return String con toda la información del jugador
     */
    public String toString(){
        String info = super.toString();
        
        //Añadimos las armas
        info += "\nWeapons:\n";
        for(int i = 0; i <weapons.size() ; i++){
            info += weapons.get(i).toString();
        }
        
        //Añadimos los escudos
        info += "\nShields:\n";
        for(int i = 0; i <shields.size() ; i++){
            info += shields.get(i).toString();
        }
        info += "\n";
        
        return info;
    }
    
    /**
     * Comprueba si algún arma se descarta, la elimina, y posteriormente le da el arma nueva al jugador
     * @param w Arma a recibir por el jugador
     */
    protected void receiveWeapon(Weapon w)
    {
        for(int i=0;i<weapons.size();i++){
            Weapon wi=weapons.get(i); 
            boolean discard=wi.discard();
            if(discard)
                weapons.remove(wi);
        }
        int size=weapons.size();
        if(size<MAX_WEAPONS)
            weapons.add(w);
    }
    
    /**
     * Comprueba si algún escudo se descarta, lo elimina, y posteriormente le da el escudo nuevo al jugador
     * @param s Escudo a recibir por el jugador
     */
    protected void receiveShield(Shield s)
    {
        for(int i=0;i<shields.size();i++){
            Shield si=shields.get(i);
            boolean discard=si.discard();
            if(discard)
                shields.remove(si);
        }
        int size=shields.size();
        if(size<MAX_SHIELDS)
            shields.add(s);
    }
    
    /**
     * Funcion newWeapon. Crea y devuelve una nueva arma con parámetros aleatorios
     * @return arma creada con parámetros aleatorios
     */
    private Weapon newWeapon(){
        return new Weapon(Dice.weaponPower(), Dice.usesLeft());
    }
    
    /**
     * Funcion newShield. Crea y devuelve un nueva escudo con parámetros aleatorios
     * @return escudo creado con parámetros aleatorios
     */
    private Shield newShield(){
        return new Shield(Dice.shieldPower() , Dice.usesLeft());
    }
    
    /**
     * Funcion sumWeapons. Devuelve el valor de ataque total de todas las armas del jugador
     * También se decrementa el uso de dichas armas (con la llamada de attack en cada arma)
     * 
     * @return float con la suma de ataque
     */
    protected float sumWeapons(){
        float totalAttack = 0f;
        for(int i = 0; i < weapons.size(); i++){
            totalAttack += weapons.get(i).attack();
        }
        return totalAttack;
    }
    
    /**
     * Funcion sumShields. Devuelve el valor de defensa total de todos los escudos del jugador
     * También se decrementa el uso de dichos escudos (con la llamada de protect en cada escudo)
     * 
     * @return float con la suma de ataque
     */
    protected float sumShields(){
        float totalDefense = 0f;
        for(int i = 0; i < shields.size(); i++){
            totalDefense += shields.get(i).protect();
        }
        return totalDefense;
    }
    
    /**
     * Funcion defensiveEnergy. Devuelve el valor de defensa del jugador, calculandolo
     * como la suma de la denfesa de sus escudos y la suma de su inteligencia
     * 
     * @return float con el valor de ataque
     */
    protected float defensiveEnergy(){
        return this.getIntelligence() + sumShields();
    }  
    
    /**
     * Administra el daño recibido por el jugador y las consecuencias del mismo.
     * - Si el daño es mayor que la suma de su inteligencia más la defensa 
     *   de sus escudos, recibe uno de daño e incrementa el numero de golpes 
     *   consecutivos recibidos
     * - Si el daño es menor que la suma de su inteligencia más la defensa 
     *   resetea los golpes consecutivos recibidos a 0
     * Después comprueba:
     *   - Si alcanza el numero de golpes consecutivos, pierde
     *   - Si tiene 0 de vida pierde
     *      
     * @param receivedAttack Valor del ataque recibido
     * @return Si el jugador ha perdido o no
     */
    private boolean manageHit(float receivedAttack){
        float defense=this.defensiveEnergy();
        if(defense < receivedAttack){
            this.gotWounded();
            this.incConsecutiveHits();
            System.out.print("\tJugador pierde un corazon\n"+consecutiveHits+" golpes consecutivos\n");

        }
        else{
            this.resetHits();
            System.out.print("\tJugador "+this.number+" se defiende\n");
        }
        
        boolean lose;
        if(consecutiveHits==HITS2LOSE){
            System.out.print("\tJugador recibe "+HITS2LOSE+" golpes, queda fuera de combate\n");
            this.resetHits();
            lose=true;
        }
        else if(this.dead()){
            System.out.print("\nJugador "+this.number+" muere\n");
            this.resetHits();
            lose=true;
        }
        else
            lose=false;
        
        return lose;
    }
    
    /**
     * Funcion resetHits. Resetea el número de golpes recibidos por el jugador
     */
    private void resetHits(){
        consecutiveHits = 0;
    }
       
    /**
     * Funcion incConsecutiveHits. Incrementa el número de golpes consecutivos
     * recibidos por el jugador
     */
    private void incConsecutiveHits(){
        consecutiveHits++;
    }
}

File: Irrgarten/lib/Player.rb
# Clase Player

# Clase Player. Representación de la clase player con sus métodos y variables 
# 
# @author Miguel Ángel Luque Gómez
# correo e.mangelluqg@go.ugr.es
  
# @author Sergio Calvo González
# correo e.sergiocg10@go.ugr.es
 
require_relative 'LabyrinthCharacter'
require_relative 'Weapon'
require_relative 'Shield'
require_relative "Directions"

class Player < LabyrinthCharacter

	public_class_method :new
	
	@@MAX_WEAPONS=2
	@@MAX_SHIELDS=3
	@@INITIAL_HEALTH=3
	@@HITS2LOSE=3	
    
    attr_accessor :number
    attr_accessor :consecutiveHits  	
	attr_accessor :weapons
	attr_accessor :shields
	
     # Constructor por parámetro de la clase Player.
     # 
     # @param nmb Número del jugador
     # @param howsmart Cómo de inteligente es
     # @param howstrong Cómo de fuerte es    
	def initialize(nmb, howSmart, howStrong)
		super("Player#"+nmb, howSmart, howStrong, @@INITIAL_HEALTH)
		@number=nmb
		@consecutiveHits=0
		@weapons=Array.new
		@shields=Array.new
	end
	
	def copy(other)
		super
		@number = other.number
		@consecutiveHits=other.consecutiveHits
		
		@weapons = other.weapons
		@shields = other.shields
			
	end
     
     # Funcion resurrect. Resucita al jugador, perdiendo las armas y escudos 
     # que poseía y recuperando la salud máxima, además de resetear el número de
     # golpes recibidos.
	def resurrect
		@weapons.clear
		@shields.clear
		@health=@@INITIAL_HEALTH
		resetHits
	end
     
     #Funcion getNumber. Devuelve el número del jugador
     #
     #@return Número del jugador 
     #
	def getNumber
		@number
	end
	
	# Funcion move. Mueve al personaje hacia direction, si no se puede, hacia una valida dentro de validMoves
	#
	# @param direction Direccion preferente a moverse
	# @param validMoves Vector con las direcciones validas a moverse
	def move(direction, validMoves)
		
		size = validMoves.size
		contained = false
		i = 0
		while ( i < size && !contained ) 
			if validMoves[i] == direction
				contained = true
			end
			i += 1
		end
		if (size > 0 && !contained)
			firstElement = validMoves[0]
			move = firstElement
		else
			move = direction
		end
		return move
	end
	
     #
     # Funcion attack. Devuelve el valor de ataque del jugador, calculandolo
     # como la suma del ataque de sus armas y la suma de su fuerza
     # 
     # @return float con el valor de ataque
     #
	def attack
		return @strength+sumWeapons
	end
     
     #
     # Funcion defend. Devuelve si el jugador se defiende del ataque
     # @param receivedAttack Cantidad de daño del atacante
     # @return True si se defiende, false si no
     #
	def defend(receivedAttack)
		return manageHit(receivedAttack)
	end
	
	# Funcion receiveReward. Le da al jugador armas, escudos y vida aleatorios
	def receiveReward
		nW = Dice.weaponsReward()
		nS = Dice.shieldsReward()
		
		nW.times do
		    w = newWeapon
			receiveWeapon(w)
		end
		
		nS.times do 
			s = newShield
			receiveShield(s)
		end
		
		extraHealth = Dice.healthReward
		@health += extraHealth
	end
	
     #
     # Funcion toString. Devuelve la información del jugador, su vida, inteligencia, fuerza
     # y la infomación de sus armas y escudos.
     # @return String con toda la información del jugador
     #
	def to_s
		info = super
		info +="\n\nWeapons:\n"
		weapons.each do |weapon| 
			info+= weapon.to_s
			info+="\n"
		end
		info +="\n\nShields:\n"
		shields.each do |shield| 
			info+= + shield.to_s
			info+="\n"
		end
		return info
	end
	
	# Funcion receiveWeapon. Elimina las armas que se deben de descartar y le da al jugador el arma nueva si se puede
	#
	# @param newW Arma nueva para el jugador
	def receiveWeapon(newW)
		@weapons.each do |w| 
			if w.discard
				@weapons.delete(w)
			end
		end
		
		if @weapons.size < @@MAX_WEAPONS
			@weapons.push(newW)
		end
	end
	
	# Funcion receiveShield. Elimina los escudos que se deben de descartar y le da al jugador el escudo nuevo si se puede
	#
	# @param newS Escudo nuevo para el jugador
	def receiveShield(newS)
		@shields.each do |s| 
			if s.discard
				@weapons.delete(s)
			end
		end
		
		if @shields.size < @@MAX_SHIELDS
			@shields.push(newS)
		end
	end
	
    #    
    # Funcion newWeapon. Crea y devuelve una nueva arma con parámetros aleatorios
    # @return arma creada con parámetros aleatorios
    #
	def newWeapon
		return Weapon.new(Dice.weaponPower, Dice.usesLeft)
	end
     
     #
     # Funcion newShield. Crea y devuelve un nueva escudo con parámetros aleatorios
     # @return escudo creado con parámetros aleatorios
     #
	def newShield
		return Shield.new(Dice.shieldPower, Dice.usesLeft)
	end

     # Funcion sumWeapons. Devuelve el valor de ataque total de todas las armas del jugador
     # También se decrementa el uso de dichas armas (con la llamada de attack en cada arma)
     # 
     # @return float con la suma de ataque
     #
	def sumWeapons
		totalAttack=0
		weapons.each do |weapon|
			totalAttack+=weapon.attack
		end
		return totalAttack
	end
     #
     # Funcion sumShields. Devuelve el valor de defensa total de todos los escudos del jugador
     # También se decrementa el uso de dichos escudos (con la llamada de protect en cada escudo)
     # 
     # @return float con la suma de ataque
     #
	def sumShields
		totalDefense=0
		shields.each do |shield|
			totalDefense+=shield.protect
		end
		return totalDefense
	end
     #
     # Funcion defensiveEnergy. Devuelve el valor de defensa del jugador, calculandolo
     # como la suma de la denfesa de sus escudos y la suma de su inteligencia
     # 
     # @return float con el valor de ataque
     #
	def defensiveEnergy
		return @intelligence + sumShields
	end
	
	#Funcion manageHit. Gestiona el ataque recibido por el jugador
	# Si el jugador recibe un ataque mayor que su inteligencia más la suma de la defensa de sus escudos, recibe uno de daño, si no, se resetea el numero de golpes consecutivos recibidos
	# Si el numeo de golpes consecutivos alcanza @@HITS2LOSE el jugador pierde
	# Si la vida del jugador es 0, el jugadir pierde tambien
	# 
	# @param receivedAttack Valor de ataque recibido
	# @return Booleano de si el jugador pierde o no
	#
	def manageHit(receivedAttack)
		defense = defensiveEnergy
				
		if defense < receivedAttack
			gotWounded
			incConsecutiveHits
		else
			resetHits
		end
		
		if (@consecutiveHits == @@HITS2LOSE) || dead
			lose = true
		else
			lose = false
		end
		
		return lose
	end
	
     #	   
     # Funcion resetHits. Resetea el número de golpes recibidos por el jugador
     #
	def resetHits
		@consecutiveHits=0
	end
     
     #
     # Funcion incConsecutiveHits. Incrementa el número de golpes consecutivos
     # recibidos por el jugador
     #
	def incConsecutiveHits
		@consecutiveHits+=1
	end
	
	protected :sumWeapons, :sumShields, :defensiveEnergy
	private :receiveWeapon, :receiveShield, :newWeapon, :newShield, :manageHit, :resetHits, :incConsecutiveHits
end
